front/lib/api/assistant/messages.ts

```diff
@@ -168,77 +170,97 @@ export async function batchRenderAgentMessages(
       visualizationActionTypesFromAgentMessageIds(agentMessageIds))(),
   ]);
 
-  return agentMessages.map((message) => {
-    if (!message.agentMessage) {
-      throw new Error(
-        "Unreachable: batchRenderAgentMessages has been filtered on agent message"
-      );
-    }
-    const agentMessage = message.agentMessage;
-
-    const actions: AgentActionType[] = [
-      agentRetrievalActions,
-      agentDustAppRunActions,
-      agentTablesQueryActions,
-      agentProcessActions,
-      agentWebsearchActions,
-      agentBrowseActions,
-      agentVisualizationActions,
-    ]
-      .flat()
-      .filter((a) => a.agentMessageId === agentMessage.id)
-      .sort((a, b) => a.step - b.step);
-
-    const agentConfiguration = agentConfigurations.find(
-      (a) => a.sId === agentMessage.agentConfigurationId
-    );
-    if (!agentConfiguration) {
-      throw new Error(
-        "Unreachable: agent configuration must be found for agent message"
+  // The only async part here is the content parsing, but it's "fake async" as the content parsing is not doing
+  // any IO or network. We need it to be async as we want to re-use the async generators for the content parsing.
+  return Promise.all(
```

Comment:
The changes here is that we move to parsing the raw contents to re-construct `content` and `chainOfThought`. The content parsing has an async signature, because we need to use the async generators from `AgentMessageContentParser`, but there is no actual IO being done.